runtime/runtime_controller.cc

```diff
@@ -121,12 +121,29 @@ bool RuntimeController::FlushRuntimeStateToIsolate() {
   FML_DCHECK(!has_flushed_runtime_state_)
       << "FlushRuntimeStateToIsolate is called more than once somehow.";
   has_flushed_runtime_state_ = true;
+
+  auto platform_configuration = GetPlatformConfigurationIfAvailable();
+  if (!platform_configuration) {
+    return false;
+  }
+
   for (auto const& [view_id, viewport_metrics] :
        platform_data_.viewport_metrics_for_views) {
-    if (!AddView(view_id, viewport_metrics)) {
+    bool added = platform_configuration->AddView(view_id, viewport_metrics);
+
+    // Callbacks will have been already invoked if the engine was restarted.
+    if (pending_add_view_callbacks_.find(view_id) !=
+        pending_add_view_callbacks_.end()) {
+      pending_add_view_callbacks_[view_id](added);
+      pending_add_view_callbacks_.erase(view_id);
+    }
+
+    if (!added) {
```

Comment:
Copying this conversation: https://github.com/flutter/engine/pull/51523#discussion_r1538200746

> @[dkwingsmt](https://github.com/dkwingsmt) (https://github.com/flutter/engine/pull/51523#discussion_r1538200746)
> We probably don't want to return false here but instead finish all other flushing. Imagine you added duplicate views by mistake. The app should still run except that the duplicate view controller is not usable.
> 
> @[loic-sharma](https://github.com/loic-sharma) (https://github.com/flutter/engine/pull/51523#discussion_r1538333859)
> My intuition agrees with you, but that's the pattern used by FlushRuntimeStateToIsolate today. For example if setting locales fails, the isolate's accessibility, settings, initial app lifecycle, and display info won't be flushed. I would punt changing this pattern to separate pull request if that's something we'd want to do.

Reply:
I don't think they're of the same case though. All current `SetXxxx` operations simply flushes the state to the platform configuration, and succeeds as long as there is a platform configuration. The `AddView`, however, also fails if the view ID is invalid, which is a user error. In short, the engine can stop initializing if the user does some wrong operation before the Dart isolate is ready, which is not something that would happen today.

Reply:
I think this code does follow today's behavior. Today `FlushRuntimeStateToIsolate` calls `AddView` for all viewport metrics. If any `AddView` returns `false`, `FlushRuntimeStateToIsolate` immediately returns `false`:

https://github.com/flutter/engine/blob/3f332b235c8973809ef8ba9114eb6055c4800bdd/runtime/runtime_controller.cc#L124-L129

Please let me know if I'm missing something.

Reply:
No. Because `AddView` may return false for two reasons:
* `add_view_.dart_state` doesn't exist, which typically happens during the destruction of the Dart isolate, and is an **expected exception**.
* The `view_id` is illegal, which is a user **error**.

All other operations, taking `SetLocales` for example, only return false if the `platform_configuration` doesn't exist, an **expected exception**. They never return false on user errors. 

In short, in today's main, the RuntimeController will never stop initializing on user errors. With this change, it will.

Reply:
Ok gotcha, will fix!

Reply:
@dkwingsmt This is fixed now!